shapely/strtree.py

```diff
@@ -209,21 +276,164 @@ def query(self, geom: BaseGeometry) -> Sequence[BaseGeometry]:
+            except ValueError:
+                raise ValueError("Could not broadcast distance to match geometry")
+
+            return self._tree.dwithin(geometry, distance)
+
+        predicate = BinaryPredicate.get_value(predicate)
+        return self._tree.query_bulk(geometry, predicate)
+
+    @requires_geos("3.6.0")
+    def _nearest_idx(self, geometry, exclusive: bool = False):
+        """Returns the index of the nearest item in the tree for each input
+        geometry.
+
+        If there are multiple equidistant or intersected geometries in the tree,
+        only a single result is returned for each input geometry, based on the
+        order that tree geometries are visited; this order may be
+        nondeterministic.
+
+        Any geometry that is None or empty in the input geometries is omitted
+        from the output.
+
+        Parameters
+        ----------
+        geometry : Geometry or array_like
+            Input geometries to query the tree.
+
+        Returns
+        -------
+        ndarray with shape (2, n)
+            The first subarray contains input geometry indexes.
+            The second subarray contains tree geometry indexes.
+
+        See also
+        --------
+        nearest_all: returns all equidistant geometries and optional distances
+
+        Examples
+        --------
+        >>> import shapely
+        >>> tree = shapely.STRtree(shapely.points(np.arange(10), np.arange(10)))
+        >>> tree.nearest(shapely.points(1,1)).tolist()  # doctest: +SKIP
+        [[0], [1]]
+        >>> tree.nearest( shapely.box(1,1,3,3)]).tolist()  # doctest: +SKIP
+        [[0], [1]]
+        >>> points = shapely.points(0.5,0.5)
+        >>> tree.nearest([None, shapely.points(10,10)]).tolist()  # doctest: +SKIP
+        [[1], [9]]
+        """
+        # TODO(shapely-2.0)
         if exclusive:
             raise NotImplementedError(
                 "The `exclusive` keyword is not yet implemented for Shapely 2.0"
             )
-        geometry = np.asarray(geom, dtype=object)
-        if geometry.ndim == 0:
-            geometry = np.expand_dims(geometry, 0)
 
-        indices = self._tree.nearest(geometry)
-        # nearest returns ndarray with shape (2, 1) -> index in input
+        geometry_arr = np.asarray(geometry, dtype=object)
+        # TODO those changes compared to _tree.nearest output should be pushed into C
+        # _tree.nearest currently ignores missing values
+        if is_missing(geometry_arr).any():
+            raise ValueError(
+                "Cannot determine nearest geometry for missing value (None)"
+            )
+        # _tree.nearest returns ndarray with shape (2, 1) -> index in input
         # geometries and index into tree geometries
-        idx = indices[1, 0]
-        return idx
+        indices = self._tree.nearest(np.atleast_1d(geometry_arr))[1]
+
+        if geometry_arr.ndim == 0:
+            return indices[0]
```

Comment:
Will this fail if `indices` is `[]`?

Reply:
A geometry always has some nearest geometry, so this should never be an empty array, except if the tree is empty. But so that gets us back to the non-inline discussion (and currently `nearest_item` will check for this and return None in that case, not getting to this line)